cpp/memilio/mobility/graph_simulation.h

```diff
@@ -97,20 +99,154 @@ class GraphSimulation
         return std::move(m_graph);
     }
 
-private:
+protected:
     double m_t;
     double m_dt;
     Graph m_graph;
     node_function m_node_func;
     edge_function m_edge_func;
 };
 
+template <class Graph>
+class GraphSimulation : public GraphSimulationBase<Graph>
+{
+    using GraphSimulationBase<Graph>::GraphSimulationBase;
+};
+
+template <class Graph>
+class GraphSimulationStochastic
+    : public GraphSimulationBase<Graph,
+                                 std::function<void(typename Graph::EdgeProperty&, size_t,
+                                                    typename Graph::NodeProperty&, typename Graph::NodeProperty&)>>
+{
+    using GraphSimulationBase<Graph,
+                              std::function<void(typename Graph::EdgeProperty&, size_t, typename Graph::NodeProperty&,
+                                                 typename Graph::NodeProperty&)>>::GraphSimulationBase;
+    using Base = GraphSimulationBase<Graph,
+                              std::function<void(typename Graph::EdgeProperty&, size_t, typename Graph::NodeProperty&,
+                                                 typename Graph::NodeProperty&)>>;
+
+public:
+    void advance(double t_max = 1.0)
+    {
+        //draw normalized waiting time
+        ScalarType normalized_waiting_time = ExponentialDistribution<ScalarType>::get_instance()(1.0);
+        std::vector<ScalarType> dt_cand(Base::m_graph.nodes().size());
+        ScalarType cumulative_rate = 0; //cumulative transition rate
+        size_t parameters_per_edge =
+            size_t(Base::m_graph.edges()[0].property.get_parameters().get_coefficients().get_shape().rows());
+        std::vector<ScalarType> transition_rates(parameters_per_edge * Base::m_graph.edges().size());
+        while (Base::m_t < t_max) {
+            Base::m_dt = std::min({Base::m_dt, t_max - Base::m_t});
+            //calculate current transition rates and cumulative rate
+            cumulative_rate = get_cumulative_transition_rate();
+            int node_iterator;
+            if (cumulative_rate * Base::m_dt >
+                normalized_waiting_time) { //at least one transition event during current time step
+                do {
+                    //draw transition event
+                    size_t event = mio::DiscreteDistribution<size_t>::get_instance()(get_rates());
+                    //edge that performs transition event
+                    auto& event_edge = Base::m_graph.edges()[int(event / parameters_per_edge)];
```

Comment:
This cast to int is superfluous, and could potentially cause an overflow. It can be safely removed.

Reply:
I think it is necessary, because 'event / parameters_per_edge' needs to be rounded down which is done with the cast to int

Reply:
It is not necessary, since both event and parameters_per_edge have type size_t  (an unsigned integer of some size), so "/" uses integer division, i.e. its result is rounded down and of type size_t. However, casting a size_t (unsigned) to int (signed) could lead to a negative / out of range index.